### ISSUE TITLE
Command line options (cli parameters)

### ISSUE THREAD

**Issue by majkinetor** [1Remote/1Remote#427]
1Remote should have several command line options that can be used for installation, startup and connection running.

1. [x] `--(un)install`
   Runs all of the (un)install-*** options
2. [x] `--(un)install-desktop-shortcut `
   (Un)Install desktop shortcut and exit
3. [x] `--(un)install-startup`
   (Un)Install registry startup entry and exit
4. [x] `--start-minimized`
   Run app minimized
5. [x] `ULID:<connection-id>` or `<connection-name>` or `#<tag-name>`
   Run app and imediatelly execute and/or focus connection name or tag. NOTE: we could have just one here if this is problematic (tag or single connection name)

## Documentation

TODO (@majkinetor)

## Examples

```
# Install everything - desktop shortcut and startup (over existing ones if found) and exit. Startup uses --miimized option
1remote.exe --install

# Remove desktop shortcut if found and exit
1remote.exe --uninstall-desktop-shortcut

# run tag #foo and connection 'bar'
1remote.exe #foo bar

#run long named connection
1remote.exe "foo bar baz" 

#run two connections by connection id
1remote.exe id:12 id:17

#run in the background
1remote.exe --start-minimized

#run and/or focus
1remote.exe
```

## Future work

After CLI options are done, here are the things that should be done in the subsequent tickets that I will create separatelly:

1. On connection context menu - add desktop shortcut
2. On connection bulk selection menu - add desktop shortcut to run selected connections
3. On tag context menu - add desktop shortcut to run this tag
4. On options menu - run automatically at OS startup (it will use --minimzed option)
    5. Start minimized should be removed as it is pointless

### Install/Uninstall

We could expand this with some more but this is enough for the start IMO and to be able to have installation routine for #402.

I intentionally didn't want to introducd option for "all users", start menu entries and uninstall in "programs and files". The first two are IMO not needed - 1RM is not a kind of program that will be used by multiple users on single machine IMO (and if that is needed, each user could have its own portable copy), and since it doesn't have multiple runable items start menu entery is usless. We could add uninstall entery that will just call `1remote.exe --uninstall` but it will leave files on the disk and it doesn't seem important. I think we could just add those in the GUI options, for example _System integration_ tab in options that would enable user to run all those things that we will be able to do via CLI (after these are implemented) via convenient buttons and/or checkboxes.

### Run minimized

Startup minimized is implemented the wrong way. 1RM should have an app argument that does that, which will then setup to be used in Windows registry. That way, when one runs 1Remote by hand or shortcut icon, it will show the main window, which is expected behavior.

**Comment by majkinetor** [comment]
There is small problem with this and winget - user will have to shutdown app first in order for it to get updated. But that is the same for all portable apps, its not specific to 1RM.

**Comment by majkinetor** [comment]
As for some examples @VShawn , check up [everything](https://www.voidtools.com/support/everything/command_line_options/)

**Comment by majkinetor** [comment]
BTW, I would also add here some command line options regarding the running of specific connections. But I consider this to be lower priority task and could be also done in future. Something like:

- `-run #tag-foo` 
- `-run 'con name 1' 

1Remote would accept multiple of those. This would enable 'create desktop shortcut` for specific connection(s).

**Comment by VShawn** [comment]
> Startup minimized is implemented the wrong way. 1RM should have an app argument that does that 

You are right :)

> `-run 'con name foo'`

I designed this feature in the prototype of PRM. users could create shortcuts in the right-click context menu of the connection. then they could search and open connections through tools like Everything. Now, we can still keep the action of creating shortcuts in context menu

P.S.

I'd like that there will be a todo list like #401, for each issue after the discussion is completed, so that we can confirm the work to be done. After all, these tickets may not be started for some time, and without a todo list, we may forget the objectives.

**Comment by majkinetor** [comment]
Yeah, you are right about todo list. However, it can be done only after the discussion, so I'll make sure we have it in the opening post.

**Comment by majkinetor** [comment]
@VShawn, I edited the first post. Please take a look and if you have comments or questions bring it on.

**Comment by VShawn** [comment]
I think it's good :)

**Comment by majkinetor** [comment]
> connection-name or #tag i think it should be case insensitive?

Yes

> And the command will start all connections named connection-name from all databases?
> from this point I wanna add a protocol like path/to/1Remote.exe id:xxxx

Well, you can have multiple same named connections even in the local database. So yes, all should be started if that is the case I guess. 

Yes, we should have connection id, but I would use it in addition to connection name as it is not CLI friendly - majority of the people won't have duplicate connection names anyway.  I will add it to the above list.

**Comment by VShawn** [comment]
I've done all the task in this ticke? Do I need to continue adding a desktop shortcut for connection in this ticket?

**Comment by majkinetor** [comment]
Lets keep this open until I test it and add documentation.

**Comment by VShawn** [comment]
auto build is donw since I import the COM `Windows Script Host Object Model` in order to creat a desktop shortcut

```
var shortcut = new IWshRuntimeLibrary.WshShell().CreateShortcut(shortcutPath);
shortcut.IconLocation =
shortcut.TargetPath = Process.GetCurrentProcess().MainModule!.FileName!;
shortcut.WorkingDirectory = Environment.CurrentDirectory;
shortcut.Arguments = "";
shortcut.Description = Assert.APP_DISPLAY_NAME;
shortcut.Save();
```

I can build it on my dev env, but if failed with github action.

Build Error from github action: `The task "ResolveComReference" is not supported on the .NET Core version of MSBuild.`B

So you are not able to test cli until I find out how to fix github action.

**Comment by majkinetor** [comment]
What about [this](https://www.nuget.org/packages/WindowsShortcutFactory/) one.

Other then that, you could embed [nircmd](https://nircmd.nirsoft.net/shortcut.html) (license allows it). Its around 100KB.

**Comment by VShawn** [comment]
It was eventually built

**Comment by majkinetor** [comment]
I tested it and everything works. Maybe we want to have `/?` as standardized but its not teribly important as we will have web docs.

**Comment by majkinetor** [comment]
Actually reopening this while docs are not done.

**Comment by majkinetor** [comment]
Ahh, this is awesome:

![image](https://github.com/1Remote/1Remote/assets/85767/ade7bb06-0562-47dc-be15-a16ab0bf6053)

**Comment by VShawn** [comment]
You even changed the icon of the shortcut. I also considered mapping the shortcut icon to the server icon when developing, but the icon of connections stored in 1RM is in PNG format while shortcuts only support ICO icons.
